### PR TITLE
[connector/spanmetrics] Produce delta temporality span metrics with timestamps representing an uninterrupted series

### LINKED ISSUE #31671: [connector/spanmetrics] Delta span metric StartTimeUnixNano doesn't follow specification
### Component(s)

connector/spanmetrics

## Expected Behaviour

The [specification](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#temporality) on metric temporality expects uninterrupted series of delta data points to have timestamps like below:

image

## Actual Behaviour

When configuring `spanmetricsconnctor` to use delta temporarily, the timestamps of successive data points follow the pattern `(T1, T1), (T2, T2), (T3, T3)`.  Basically the `StartTimeUnixNano` and `TimeUnixNano` are always set to the current timestamp.

When configured to use cumulative temporality, the connector caches `StartTimeUnixNano` so that it can be used in the next data point. However, when using delta temporality that cache (`resourceMetrics`) gets wiped after each round of exporting metrics:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/71e6c553f73feb6ea99f3f9d625c42e2267a8ec3/connector/spanmetricsconnector/connector.go#L283-L301

### Collector version

v0.94.0

### Environment information

Testing done locally on mac OS

### OpenTelemetry Collector configuration

```yaml
connectors:
  spanmetrics:
    dimensions_cache_size: 5000
    resource_metrics_key_attributes:
      - service.name
    aggregation_temporality: "AGGREGATION_TEMPORALITY_DELTA"
    metrics_flush_interval: 5s
    exemplars:
      enabled: true
    histogram:
      explicit:
        buckets: ...

exporters:
  prometheus/exporter:
    endpoint: "0.0.0.0:9465"
    enable_open_metrics: true

service:
  pipelines:
    traces:
      receivers: [otlp]
      processors: []
      exporters: [spanmetrics]
    metrics:
      receivers: [otlp, spanmetrics]
      processors: []
      exporters: [prometheus/exporter]
```

### Log output

_No response_

## Why fix this? 
This could be a different way of fixing https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30688 and https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30559. It could also significantly reduce memory usage when combining `spanmetricsconnector` with exporters like `prometheusexporter`.

## Issue with Prometheus Exporter

`prometheusexporter` supports conversion from delta to cumulative temporality. However, it only works if the series follows the specification:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/71e6c553f73feb6ea99f3f9d625c42e2267a8ec3/exporter/prometheusexporter/accumulator.go#L203-L212

I think the author's rationale is the delta series seems reset based on [this part of the spec](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#resets-and-gaps). So with the current delta span metric behaviour, prometheus views each data point as if the counter was reset. 

A workaround is having the connector produce cumulative span metrics, but that requires the connector caching all series in memory, which is redundant since `prometheusexporter` already does this. @matej-g is working on an improvement to expire infrequently updated series from the cache which will definitely help with reducing memory.

## Solution Discussion

Having delta span metrics follow the spec could significantly reduce memory usage when used with `prometheusexporter`.

To do this, I think the connector could probably have to cache the `StartTimeUnixNano` for each series when in delta mode. The currently cached value for each series are much larger since they include things like metric attributes and histogram bucket sizes

### PR DESCRIPTION
**Pull request by swar8080**
Closes #31671

**Description:** 
Currently delta temporality span metrics are produced with (StartTimestamp, Timestamp)'s of `(T1, T2), (T3, T4) ...`. However, the [specification](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#temporality) says that the correct pattern for an uninterrupted delta series is `(T1, T2), (T2, T3) ...`

This misalignment with the spec can confuse downstream components' conversion from delta temporality to cumulative temporality, causing each data point to be viewed as a cumulative counter "reset". An example of this is in `prometheusexporter`

The conversion issue forces you to generate cumulative span metrics, which use significantly more memory to cache the cumulative counts.

At work, I applied this patch to our collectors and switched to producing delta temporality metrics for `prometheusexporter` to then convert to cumulative. That caused a significant drop in-memory usage:

image

**Testing:** 
- Unit tests asserting the timestamps
- Manual testing with `prometheusexporter` to make sure counter values are cumulative and no longer being reset after receiving each delta data point

### CODE REVIEW

File: connector/spanmetricsconnector/README.md
Comment (swar8080) [inline connector/spanmetricsconnector/README.md:128]: Thoughts on this default? Existing users of this component will have higher memory usage

Also, I can open another task to add a separate section to the README to better explain how using delta temporality + `prometheusexporter` can reduce memory

File: connector/spanmetricsconnector/README.md
Comment (swar8080) [inline connector/spanmetricsconnector/README.md:127]: Should I open another task that could make the configuration easier to use? The changes I have in mind are:

1. Deprecate `aggregation_temporality` and remove it from the documentation

2. Add `cumulative_temporality` to the configuration with an `enabled` property. Also move `metrics_expiration` as a property, since it's exclusively used for cumulative temporality

3. Add an `enabled` property to `delta_temporality`

4. Update `config.go` to honour both the new and old approach for picking temporality, and validate that there's no ambiguity about which temporality to use

File: connector/spanmetricsconnector/connector.go
Comment (swar8080) [inline connector/spanmetricsconnector/connector.go:364]: Added this just to make testing easier, since before that the `startTimestamp` was regenerated during each iteration of the for loop on the next line

**Comment by djaglowski** [comment]
Thanks for the issue and PR @swar8080. Hopefully @portertech or someone from @open-telemetry/collector-contrib-approvers can review.

File: connector/spanmetricsconnector/README.md
Comment (Frapschen) [reply connector/spanmetricsconnector/README.md:127]: I would rather not introduce breaking change, like https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31106. 

we just add a new config option `metric_timestamp_cache_size` and give a proper default value

File: connector/spanmetricsconnector/connector.go
Comment (Frapschen) [inline connector/spanmetricsconnector/connector.go:284]: Is there any reason we use an anonymous function?  I suggest use naming function.

I think we can get the cached startTime before `resourceMetrics.ForEach` and just use it in the loop.

File: connector/spanmetricsconnector/connector.go
Comment (Frapschen) [inline connector/spanmetricsconnector/connector.go:291]: Can we update `lastDeltaTimestamps` here directly?

Comment (swar8080) [reply connector/spanmetricsconnector/connector.go:291]: This anonymous function is first called for the `calls` metric and then called with the same key for the `duration` metric. So if we update it here then the `duration` metric will get the incorrect value from the cache

That is what I was trying to say in the comment above this line, but feel free to change the wording if it's unclear!

File: connector/spanmetricsconnector/README.md
Comment (swar8080) [reply connector/spanmetricsconnector/README.md:127]: Okay I can make that change

I was thinking it might be confusing if some configuration options are used for one temporality but not the other

File: connector/spanmetricsconnector/README.md
Comment (swar8080) [reply connector/spanmetricsconnector/README.md:128]: Agreed. What do you think the default `metric_timestamp_cache_size` should be? Memory usage will now be higher when using `AGGREGATION_TEMPORALITY_DELTA` in order to cache the keys and timestamps

Comment (swar8080) [reply connector/spanmetricsconnector/README.md:128]: > think it's great that we have a separate section to explain how to reduce memory in that use case.
sounds good, I may write that guide in a different pull request

File: connector/spanmetricsconnector/connector.go
Comment (Frapschen) [inline connector/spanmetricsconnector/connector.go:134]: Can we use `cache.NewCache` instead? This will help us maintain uniformity.

Comment (Frapschen) [inline connector/spanmetricsconnector/connector.go:134]: Can we use `cache.NewCache` instead? This will help us maintain uniformity.

File: connector/spanmetricsconnector/README.md
Comment (Frapschen) [reply connector/spanmetricsconnector/README.md:128]: > metric_timestamp_cache_size

How about as same as `defaultResourceMetricsCacheSize` 1000?

File: connector/spanmetricsconnector/connector.go
Comment (Frapschen) [reply connector/spanmetricsconnector/connector.go:284]: how about this :
```go
	p.resourceMetrics.ForEach(func(_ resourceKey, rawMetrics *resourceMetrics) {
		rm := m.ResourceMetrics().AppendEmpty()
		rawMetrics.attributes.CopyTo(rm.Resource().Attributes())

		sm := rm.ScopeMetrics().AppendEmpty()
		sm.Scope().SetName("spanmetricsconnector")

		/**
		 * To represent an uninterrupted stream of metrics as per the spec, the (StartTimestamp, Timestamp)'s of successive data points should be:
		 * - For cumulative metrics: (T1, T2), (T1, T3), (T1, T4) ...
		 * - For delta metrics: (T1, T2), (T2, T3), (T3, T4) ...
		 */
		startTimetamp := rawMetrics.startTimestamp
		if p.config.GetAggregationTemporality() == pmetric.AggregationTemporalityDelta {
			if lastTimestamp, ok := p.lastDeltaTimestamps.Get(mk); ok {
				startTime = lastTimestamp
			}

			// For delta metrics, cache the current data point's timestamp, which will be the start timestamp for the next data points in the series
			p.lastDeltaTimestamps.Add(mk, timestamp)
		}

		sums := rawMetrics.sums
		metric := sm.Metrics().AppendEmpty()
		metric.SetName(buildMetricName(p.config.Namespace, metricNameCalls))
		sums.BuildMetrics(metric, startTimetamp, timestamp, p.config.GetAggregationTemporality())
		if !p.config.Histogram.Disable {
			histograms := rawMetrics.histograms
			metric = sm.Metrics().AppendEmpty()
			metric.SetName(buildMetricName(p.config.Namespace, metricNameDuration))
			metric.SetUnit(p.config.Histogram.Unit.String())
			histograms.BuildMetrics(metric, startTimetamp, timestamp, p.config.GetAggregationTemporality())
		}

		events := rawMetrics.events
		if p.events.Enabled {
			metric = sm.Metrics().AppendEmpty()
			metric.SetName(buildMetricName(p.config.Namespace, metricNameEvents))
			events.BuildMetrics(metric, startTimetamp, timestamp, p.config.GetAggregationTemporality())
		}
	})
```

Comment (swar8080) [reply connector/spanmetricsconnector/connector.go:284]: That would be easier to understand but it doesn't work for the reason mentioned here https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31780#discussion_r1555124674

File: connector/spanmetricsconnector/connector.go
Comment (swar8080) [reply connector/spanmetricsconnector/connector.go:134]: We could but `cache.NewCache` holds onto items evicted from its LRU in memory, which isn't needed like it is for the other caches. We would then have to remember to call `RemoveEvictedItems()` to discard those items from memory

File: connector/spanmetricsconnector/README.md
Comment (swar8080) [reply connector/spanmetricsconnector/README.md]: made the change

**Comment by swar8080** [comment]
Hi @Frapschen @portertech is this something you still want to include? The component should be usable with `prometheusexporter` now that I fixed https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32210

So this MR is mostly to allow using delta metrics as an optimization. It reduces memory/CPU by not buffering the cumulative counts since `prometheusexporter` already does this, and only pushes metrics that changed to the next component.

Also having timestamps aligned with the specification might prevent future bug reports. For example, i'm still maintaining a fork with this timestamp change in order to work with https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32521 (which is also forked).

**Review by ankitpatel96 — APPROVED** [review]
This is a great change to fix a nasty bug!  I'm not an approver but I think this is in a mergeable state

**Comment by mx-psi** [comment]
@swar8080 would you have time to fix the merge conflicts?

File: connector/spanmetricsconnector/config.go
Comment (Frapschen) [inline connector/spanmetricsconnector/config.go:117]: remove this struct since not used

**Review by Frapschen — APPROVED** [review]

File: connector/spanmetricsconnector/connector_test.go
Comment (swar8080) [reply connector/spanmetricsconnector/connector_test.go:1738]: that is clearer, thanks

**Comment by swar8080** [comment]
Thanks for reviewing. I've resolved the merge conflicts and implemented the latest round of review suggestions. Should be good to merge once the pipeline passes @mx-psi

File: exporter/elasticsearchexporter/integrationtest/go.mod
Comment (swar8080) [inline exporter/elasticsearchexporter/integrationtest/go.mod:64]: needed because this module depends on spanmetricsconnector

**Comment by swar8080** [comment]
Finished implementing the most recent review feedback so this seems ready to merge cc @mx-psi

**Review by mx-psi — APPROVED** [review]

**Pull request merged**
